### ISSUE TITLE
case study on SPF

### ISSUE THREAD

**Issue by hdatma** [trusteddomainproject/OpenDMARC#180]
This is a puzzle. Enjoy it.

This is from the log:

> [...] : < o4.email.wetransfer.com[**192.254.123.89**]:22194: EHLO o4.email.wetransfer.com
> [...]: < o4.email.wetransfer.com[192.254.123.89]:22194: MAIL FROM:<bounces+922094-4260-username=example.com@**em9713.email.wetransfer.com**>

This is from the e-mail:

> X-Envelope-From: <bounces+922094-4260-username=example.com@em9713.email.wetransfer.com>
> Authentication-Results: example.com; **dmarc=fail** (p=quarantine dis=none) header.from=wetransfer.com
> Authentication-Results: example.com; **spf=tempfail** smtp.mailfrom=bounces+922094-4260-username=example.com@**em9713.email.wetransfer.com**
> From: WeTransfer <noreply@wetransfer.com>
> Message-ID: <[anonymized]@wetransfer.com>
> To: username@example.com [anonymized]

The Authentication Results were generated by DMARC via libspf2. 
Wetransfer uses DKIM as well, but I disabled this test temporarily.

Since I cherish DMARC, the fact that dmarc=fail is sufficient to put this e-mail in the junk. 

However, the e-mail is legitimate. 

So, why did DMARC fail? 

Since I disabled DKIM, then DMARC failed because spf=tempfail. 

However, if you check SPF by hand, this is what you get.

> $ unbound-host -rvD -tTXT **em9713.email.wetransfer.com**
> em9713.email.wetransfer.com is an alias for u922094.wl095.sendgrid.net. (insecure)
u922094.wl095.sendgrid.net has TXT record "v=spf1 ip4:149.72.250.193 ip4:167.89.35.243 ip4:167.89.35.32 ip4:167.89.5.15 ip4:167.89.54.234 ip4:167.89.54.238 ip4:167.89.54.247 ip4:192.254.117.71 ip4:192.254.118.54 ip4:192.254.123.42 ip4:**192.254.123.89** -all" (insecure)

So, the IP 192.254.123.89 is allowed to send on behalf of wetransfer.com, because it appears in the SPF policy at em9713.email.wetransfer.com.

So, why is spf=tempfail in the Authentication-Results?

Oh, you say, it is TEMPFAIL. Then why is that all WeTransfer e-mails today were spf=tempfail by DMARC and spf=pass by hand?

I do not have the answer. Something is wrong inside the DMARC-SPF code.

**Comment by Swallowtail23** [comment]
Interesting puzzle! 

First question is have you disabled DKIM **only** to test this SPF issue? OpenDMARC README file, "Dependencies":

Actual use requires... **some DKIM checking service that acts on messages before they reach theopendmarc milter (such as the Trusted Domain Project's "OpenDKIM").** 

Back to the SPF question :)

I wonder if the code is only doing 10 evaluations even though the RFC does not limit evaluation of ip4 items to 10, only DNS lookups. That one is the 11th ip4. Maybe coincidence though. Are all the failed wetransfer emails from the same 11th IP?

> dmarc=fail is sufficient to put this e-mail in the junk.

A comment on this. Stats from my server show that consistently about 7% of inbound HAM email fails DMARC (with p=none, not quarantine or reject). Automatically rejecting all DMARC=fail will absolutely result in false positives.

**Comment by hdatma** [comment]
Yes, there is a problem with long SPF policies that I have noted also on other domains.

Today I spotted a domain whose DMARC's SPF fails while succeeding by hand as well as by libspf2. Its policy includes two CIDRs, and it failed on the first one. Since straight libspf2 passes the test, the problem is inside DMARC.

* `spfquery -ip [sending-ip] -sender postmaster@[envelope-from]`

**Comment by thegushi** [comment]
@hdatma , I apologize that it's been a while since you put this request in, but are you using opendmarc's internal spf checking, or linked against libspf2?

**Comment by thegushi** [comment]
Thanks for the detail. Since you confirmed this is using libspf2, a transient DNS failure during the CNAME resolution (em9713.email.wetransfer.com → u922094.wl095.sendgrid.net) would be the most likely explanation for a tempfail — libspf2 would surface that as spf=tempfail rather than retrying silently.

Given this is from 2021 and SendGrid/WeTransfer SPF records have almost certainly changed since then, is this still reproducible? If so, any chance of capturing the DNS query/response at the time of the failure to confirm whether it's a resolver issue or something inside libspf2?

**Comment by hdatma** [comment]
Too much water under the bridge. Nobody can reproduce the error at this time. Please close the ticket.
